app/test/vesting_test.go

```diff
@@ -0,0 +1,498 @@
+		assert.NoError(s.T(), err)
+
+		return gs
+	}
+
+	cctx, _, _ := s.NewNetworkWithGenesisOpt(kr, gsOpt)
+
+	assert.NoError(s.T(), cctx.WaitForNextBlock())
+
+	for _, tt := range tests {
+		s.Run(tt.name, func() {
+			accAddress := getAddress(tt.accName, cctx.Keyring).String()
+
+			// Test account details correctness
+			cli := authtypes.NewQueryClient(cctx.GRPCClient)
+			res, err := cli.Account(context.Background(), &authtypes.QueryAccountRequest{
+				Address: accAddress,
+			})
+			assert.NoError(s.T(), err)
+
+			var qAcc vestingtypes.DelayedVestingAccount
+			err = qAcc.Unmarshal(res.Account.Value)
+			assert.NoError(s.T(), err)
+
+			// Checking the queried account data
+			assert.Equal(s.T(), accAddress, qAcc.GetAddress().String(), "account address must match")
+			assert.EqualValues(s.T(), tt.endTime.Unix(), qAcc.GetEndTime(), "end time must match")
+
+			assert.EqualValues(s.T(),
+				testfactory.BaseAccountDefaultBalance,
+				qAcc.GetOriginalVesting().AmountOf(app.BondDenom).Int64(),
+				"original vesting must match")
+
+			/*--------*/
+
+			// Test the locking mechanism
+			// If the end time is already passed, the funds must be unlocked
+			// and we should be able to transfer some of it to another account
+			mustSucceed := tt.endTime.Before(time.Now())
+
+			assert.NoError(s.T(), cctx.WaitForNextBlock())
+
+			/*--------*/
+
+			// Test available balance
+			balances, err := GetAccountSpendableBalance(cctx.GRPCClient, accAddress)
+			assert.NoError(s.T(), err)
+
+			expectedSpendableBal := int64(0)
+			if mustSucceed {
+				expectedSpendableBal = testfactory.BaseAccountDefaultBalance
+			}
+			expectedSpendableBal += initBalanceForGasFee.Amount.Int64()
+			assert.EqualValues(s.T(),
+				expectedSpendableBal,
+				balances.AmountOf(app.BondDenom).Int64(),
+				"spendable balance must match")
+
+			/*--------*/
+
+			s.testTransferVestingAmount(cctx, tt.accName, mustSucceed)
+
+			balancesAfter, err := GetAccountSpendableBalance(cctx.GRPCClient, accAddress)
+			assert.NoError(s.T(), err)
+
+			expectedSpendableBalAfterTx := expectedSpendableBal - 1 // -1utia for gas fee of the tx above
+			if mustSucceed {
+				// if the transfer was successful
+				expectedSpendableBalAfterTx -= testfactory.BaseAccountDefaultBalance
+			}
+
+			assert.EqualValues(s.T(),
+				expectedSpendableBalAfterTx,
+				balancesAfter.AmountOf(app.BondDenom).Int64(),
+				"spendable balance must be equal")
+
+			/*--------*/
+
+			// test delegation of only the locked account(s)
+			if !mustSucceed {
+				s.testDelegation(cctx, tt.accName)
+			}
+		})
+	}
+}
+
+func (s *VestingModuleTestSuite) TestGenesisPeriodicVestingAccounts() {
+	// initial unlocked allocation to pay the gas fees
+	initBalanceForGasFee := sdk.NewCoin(app.BondDenom, sdk.NewInt(10))
+
+	startTime := tmtime.Now()
+	periods := vestingtypes.Periods{
+		vestingtypes.Period{Length: int64(10), Amount: sdk.Coins{sdk.NewInt64Coin(app.BondDenom, 2500)}},
```

Comment:
Q: Is the length here in terms of blocks or time?

Reply:
it is time in seconds